rfc/src/rfcs/0009-function-contracts.md

```diff
@@ -0,0 +1,845 @@
+  - **The bold** (or naïve) user can skip all contracts with `--all-trusted`.
+  - **For the lawyer** that is only interested in checking contracts and nothing
+    else a `--litigate` flag checks only contract harnesses.
+
+  Aside: I'm obviously having some fun here with the names, happy to change,
+  it's really just about the semantics.
+- **Can `old` accidentally break scope?** The `old` function cannot reference local
+  variables. For instance `#[ensures({let x = ...; old(x)})]` cannot work as an
+  AST rewrite because the expression in `old` is lifted out of it's context into
+  one where the only bound variables are the function arguments (see also
+  [history variables](#history-variables)). In most cases this will be a
+  compiler error complaining that `x` is unbound, however it is possible that
+  *if* there is also a function argument `x`, then it may silently succeed the
+  code generation but confusingly fail verification. For instance `#[ensures({
+  let x = 1; old(x) == x })]` on a function that has an argument named `x` would
+  *not* hold.
+
+  To handle this correctly we would need an extra check that detects if `old`
+  references local variables. That would also enable us to provide a better
+  error message than the default "cannot find value `x` in this scope".
+- **Can panicking be expected behavior?** Usually preconditions are used to rule
+  out panics but it is conceivable that a user would want to specify that a
+  function panics under certain conditions. Specifying this would require an
+  extension to the current interface.
+- **UB checking.** With unsafe rust it is possible to break the type system
+  guarantees in Rust without causing immediate errors. Contracts must be
+  cognizant of this and enforce the guarantees as part of the contract *or*
+  require users to explicitly defer such checks to use sites. The latter case
+  requires dedicated support because the potential UB must be reflected in the
+  havoc.
+- **`modifies` clauses over patterns.** Modifies clauses mention values bound in
+  the function header and as a user I would expect that if I use a pattern in
+  the function header then I can use the names bound in that pattern as lvalues
+  in the `modifies` clause. However `modifies` clauses are implemented as
+  `assigns` clauses in CBMC which does not have a notion of function header
+  patterns. Thus it is necessary to project any `modifies` ranges deeper by the
+  fields used in the matched pattern.
+
+<!-- 
+- Is there any part of the design that you expect to resolve through the RFC process?
+- What kind of user feedback do you expect to gather before stabilization? How will this impact your design? 
+-->
+
+## Future possibilities
+
+<!-- For Developers -->
+
+- **Quantifiers:** Quantifiers are like logic-level loops and a powerful
+  reasoning helper. CBMC has support for both `exists` and `forall`, but the
+  code generation is difficult. The most ergonomic and easy way to implement
```

Comment:
What is difficult ? The code generation of quantifiers from MIR to CBMC or the handling of quantifiers in CBMC ?

Reply:
The next few sentences are supposed to explain that. Are they not comprehensible?

Reply:
maybe an example would help to illustrate the gap in representation